### PR TITLE
Avoid extra memory copies during mirror scrub.

### PR DESCRIPTION
**Pull request by amotin**
When several mirror scrub reads complete successfully, we do not need
to copy data from each of them into the parent ZIO buffer, the first
one is enough.  This also moves the memory copy out of the ZIO locks,
that I expected to cause contention, though on my system they almost
didn't, had incorrect lock order and was partially not needed.

When scrubbing 4-way mirror with 1.5TB of 128KB ZVOL blocks, this
reduces CPU time spent in memcpy() by 75% and total CPU time by 31%.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Review by behlendorf — APPROVED** [review]
Looks good.

File: module/zfs/vdev_mirror.c [hunk 1]
```diff
@@ -451,17 +452,17 @@ vdev_mirror_scrub_done(zio_t *zio)
...
 
 	if (zio->io_error == 0) {
-		zio_t *pio;
-		zio_link_t *zl = NULL;
-
-		mutex_enter(&zio->io_lock);
-		while ((pio = zio_walk_parents(zio, &zl)) != NULL) {
-			mutex_enter(&pio->io_lock);
-			ASSERT3U(zio->io_size, >=, pio->io_size);
+		zio_t *pio = zio_unique_parent(zio);
```
Comment (mmaybee) [inline module/zfs/vdev_mirror.c:455]: The original algorithm walked all parents to do the copy... while it seems reasonable that there should generally be only a single parent here, do you have a reason to believe that is an invariant? Why did the original algorithm assume the possibility of multiple parents?

**Review by mmaybee — COMMENTED** [review]
Related to this change: it looks like the mirror scrub code is only going to check a single "random" child. Whichever child gets its data into the parent io_abd "wins". Could we compare the other children's buffer contents against the winner's to make sure that all children have the same data?

File: module/zfs/vdev_mirror.c (hunk 1 above)
Comment (amotin) [reply module/zfs/vdev_mirror.c:455]: I have no idea what the original author was thinking about.  zio_unique_parent() actually verifies there are not other parents, and the assertion didn't fire on my tests.  My best guess is that it was some paranoia, same as unneeded locking.

**Pull request closed without merging**